test/sasl.ts

```diff
@@ -51,6 +51,59 @@ describe('sasl plain', function() {
     });
 });
 
+describe('sasl plain callback using promise', function() {
+    this.slow(200);
+    var container: rhea.Container, listener: any;
+    var providerFailure : boolean;
+
+    function authenticate(username: string, password: string) {
+        return new Promise((resolve, reject) => {
+            if (providerFailure) {
+                reject("provider failure");
+            } else {
+                resolve(username.split("").reverse().join("") === password);
```

Comment:
I'd suggest making this actually asynchronous rather than just returning a resolved promise. E.g. the notification of desire to write is needed for an asynch call but not if a resolved promise is returned, and that is the sort of thing that can get broken.

So e.g. process.nextTick(function () {
            if (providerFailure) {
                reject("provider failure");
            } else {
                resolve(username.split("").reverse().join("") === password);
            }
});

Wdyt?

Reply:
Sounds very wise. Done.